### PR TITLE
Refactor components to not use unsafe react methods

### PR DESCRIPTION
**Pull request by PatoBeltran**
Balloon, number and date components were using unsafe methods... This PR aims to remove them by moving the logic they had to use hooks. This is a step forward towards react concurrent mode.

### CODE REVIEW

File: lib/components/DateTime/Calendar.tsx [hunk 1]
```diff
@@ -316,132 +239,165 @@ export class Calendar extends React.Component<CalendarProps, Partial<CalendarSta
...
+                    <Attr.button
+                        type='button'
+                        aria-label={ariaLabel}
+                        data-row={rowIndex}
+                        data-col={colIndex}
+                        onKeyDown={onKeyDown}
+                        className={colClassName}
+                        onClick={handleClick}
+                        key={key}
+                        attr={props.attr?.dateButton}
```
Comment (vishwam) [inline lib/components/DateTime/Calendar.tsx]: nit: do `const { dateButton, dateRow... } = props.attr ?? {}` so we don't do the null check on attr multiple times?

File: lib/components/DateTime/DateField.tsx [hunk 2]
```diff
@@ -144,7 +141,6 @@ DateField.defaultProps = {
     label: undefined,
     onChange: undefined,
     format: DateFormat.MMDDYYYY,
-    tabIndex: -1,
```
Comment (vishwam) [inline lib/components/DateTime/DateField.tsx]: why is this removed?

File: lib/components/DateTime/DatePicker.tsx [hunk 3]
```diff
@@ -170,293 +109,267 @@ export class DatePicker extends React.Component<DatePickerProps, Partial<DatePic
...
-                parsed.month - 1,
-                parsed.date
-            ).dateObject.toJSON() : null;
-
-        return (
+        return () => {
+            window.removeEventListener('click', onOuterMouseEvent);
+            window.removeEventListener('keydown', onKeydown);
+        };
+    }, []);
```
Comment (vishwam) [inline lib/components/DateTime/DatePicker.tsx]: add dependency on onKeydown and onOuterMouseEvent?

File: lib/components/DateTime/TimeInput.tsx [hunk 4]
```diff
@@ -254,6 +201,22 @@ export class TimeInput extends React.Component<TimeInputProps, TimeInputState> {
         const seconds = this.state.seconds < 10 ? `0${this.state.seconds}` : this.state.seconds;
         const inputClassName = css('time-input', {'error': this.props.error}, this.props.inputClassName);
 
+        const hoursList = [];
```
Comment (vincenzocaruso) [inline lib/components/DateTime/TimeInput.tsx]: why did you move on the render rather then having it created one time?

File: lib/components/DateTime/TimeInput.tsx [hunk 5]
```diff
@@ -254,6 +201,22 @@ export class TimeInput extends React.Component<TimeInputProps, TimeInputState> {
...
+        const minutesList = [];
+        const secondsList = [];
+
+        const numHours = this.props.militaryTime ? 24 : 12;
+        for (let index = 0; index < numHours; index++) {
+            const value = index < 10 ? `0${index}` : `${index}`;
+            hoursList.push({label: value, value: value});
```
Comment (vishwam) [inline lib/components/DateTime/TimeInput.tsx]: instead of creating the intermediate hoursList array, store the jsx elements directly ` optionMap(option, this.props.attr.hourOption)`. That way, we'll preserve existing behavior of only looping over the hours once.

File: lib/components/Input/NumberInput.tsx [hunk 6]
```diff
@@ -226,12 +226,6 @@ export class NumberInput extends React.Component<NumberInputProps, NumberInputSt
         }
     }
 
-    UNSAFE_componentWillReceiveProps(newProps: NumberInputProps) {
```
Comment (vishwam) [inline lib/components/Input/NumberInput.tsx]: don't we need the getDerivedStateFromProps here?

File: lib/components/DateTime/TimeInput.tsx (hunk 4 above)
Comment (vishwam) [reply lib/components/DateTime/TimeInput.tsx]: or, instead of creating the intermediate hoursList array, store the jsx elements directly: optionMap(option, this.props.attr.hourOption). That way, we'll preserve existing behavior of only looping over the hours once.

File: lib/components/Input/NumberInput.tsx (hunk 6 above)
Comment (PatoBeltran) [reply lib/components/Input/NumberInput.tsx]: No, the numberInput already manages it's own state... This cwrp is allowing the user to update the initialValue without user interaction (i.e. the user changes another field and the number field is updated).

With this change that scenario won't be supported anymore, but as far as I understand we don't do that type of behavior anywhere, so this change should be safe. Another thing I could do is move this logic to componentDidUpdate and update the state there, but this would be updated a rerender later than how it use to. 

Which solution do you think is better?

File: lib/components/DateTime/TimeInput.tsx (hunk 4 above)
Comment (PatoBeltran) [reply lib/components/DateTime/TimeInput.tsx]: I did the approach @vishwam suggested

File: lib/components/Input/NumberInput.tsx (hunk 6 above)
Comment (PatoBeltran) [reply lib/components/Input/NumberInput.tsx]: Moved it to componentDidUpdate
